pkg/reporter/github_reporter.go

```diff
@@ -0,0 +1,322 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sethvargo/go-githubactions"
+
+	"github.com/abcxyz/guardian/pkg/flags"
+	"github.com/abcxyz/guardian/pkg/github"
+	"github.com/abcxyz/pkg/logging"
+)
+
+var _ Reporter = (*GitHubReporter)(nil)
+
+const (
+	githubCommentPrefix        = "#### 🔱 Guardian 🔱"
+	githubMaxCommentLength     = 65536
+	githubDestroyIndicatorText = "💥 DESTROY"
+	githubTruncatedMessage     = "\n\n> Message has been truncated. See workflow logs to view the full message."
+)
+
+var githubOperationText = map[Operation]string{
+	PlanOperation:    "PLAN",
+	ApplyOperation:   "APPLY",
+	UnknownOperation: "UNKNOWN",
+}
+
+var githubStatusText = map[Status]string{
+	StartStatus:     "🟨 STARTED",
+	SuccessStatus:   "🟩 SUCCESS",
+	NoChangesStatus: "🟦 NO CHANGES",
+	FailureStatus:   "🟥 FAILED",
+	UnknownStatus:   "⛔️ UNKNOWN",
+}
+
+var (
+	tildeChanged = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			"([~])") // tilde represents changes and needs switched to exclamation for git diff
+
+	swapLeadingWhitespace = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			`((\-(\/\+)*)|(\+(\/\-)*)|(!))`) // match characters to swap whitespace for git diff (+, +/-, -, -/+, !)
+)
+
+// GitHubReporter implements the reporter interface for writing GitHub comments.
+type GitHubReporter struct {
+	gitHubClient github.GitHub
+	flags        flags.GitHubFlags
+	commentID    int64
+	logURL       string
+}
+
+// NewGitHubReporter creates a new GitHubReporter.
+func NewGitHubReporter(ctx context.Context, f flags.GitHubFlags) (Reporter, error) {
```

Comment:
It feels weird to couple this to "flags"; these are really inputs to the reporter, right?

Reply:
Changed to inputs